libraries/chain/sps_objects/sps_processor.cpp

```diff
@@ -0,0 +1,286 @@
+
+namespace steem { namespace chain {
+
+using steem::protocol::asset;
+using steem::protocol::operation;
+
+using steem::chain::proposal_object;
+using steem::chain::by_start_date;
+using steem::chain::by_end_date;
+using steem::chain::proposal_index;
+using steem::chain::proposal_id_type;
+using steem::chain::proposal_vote_index;
+using steem::chain::by_proposal_voter;
+using steem::chain::by_voter_proposal;
+using steem::protocol::proposal_pay_operation;
+using steem::chain::sps_helper;
+using steem::chain::dynamic_global_property_object;
+using steem::chain::block_notification;
+
+const std::string sps_processor::removing_name = "sps_processor_remove";
+const std::string sps_processor::calculating_name = "sps_processor_calculate";
+
+bool sps_processor::is_maintenance_period( const time_point_sec& head_time ) const
+{
+   return db.get_dynamic_global_properties().next_maintenance_time <= head_time;
+}
+
+void sps_processor::remove_proposals( const time_point_sec& head_time )
+{
+   FC_TODO("implement proposal removal based on automatic actions")
+   auto& proposalIndex = db.get_mutable_index< proposal_index >();
+   auto& byEndDateIdx = proposalIndex.indices().get< by_end_date >();
+
+   auto& votesIndex = db.get_mutable_index< proposal_vote_index >();
+   auto& byVoterIdx = votesIndex.indices().get< by_proposal_voter >();
+
+   auto found = byEndDateIdx.upper_bound( head_time );
+   auto itr = byEndDateIdx.begin();
+
+   sps_removing_reducer obj_perf( db.get_sps_remove_threshold() );
+
+   while( itr != found )
+   {
+      itr = sps_helper::remove_proposal< by_end_date >( itr, proposalIndex, votesIndex, byVoterIdx, obj_perf );
+      if( obj_perf.done )
+         break;
+   }
+}
+
+void sps_processor::find_active_proposals( const time_point_sec& head_time, t_proposals& proposals )
+{
+   const auto& pidx = db.get_index< proposal_index >().indices().get< by_start_date >();
+
+   std::for_each( pidx.begin(), pidx.upper_bound( head_time ), [&]( auto& proposal )
+                                             {
+                                                if( head_time >= proposal.start_date && head_time <= proposal.end_date )
+                                                   proposals.emplace_back( proposal );                                                
+                                             } );
+}
+
+uint64_t sps_processor::calculate_votes( const proposal_id_type& id )
+{
+   uint64_t ret = 0;
+
+   const auto& pvidx = db.get_index< proposal_vote_index >().indices().get< by_proposal_voter >();
+   auto found = pvidx.find( id );
+
+   while( found != pvidx.end() && found->proposal_id == id )
+   {
+      const auto& _voter = db.get_account( found->voter );
+
+      //If _voter has set proxy, then his votes aren't taken into consideration
+      if( _voter.proxy == STEEM_PROXY_TO_SELF_ACCOUNT )
+      {
+         auto sum = std::accumulate(   _voter.proxied_vsf_votes.begin(),
```

Comment:
Should probably just call `_voter.proxied_vsf_votes_total()` to avoid repeated code.